Prose/ProseLib/Sources/JoinChatFeature/AddMemberSheet.swift

```diff
@@ -0,0 +1,256 @@
+      }
+      .controlSize(.large)
+      .padding(24)
+      .frame(width: 360)
+    }
+  }
+
+  @ViewBuilder
+  static func infoMessage(for info: ViewState.InfoMessage) -> some View {
+    switch info {
+    case .none, .invalid:
+      EmptyView()
+    case .unknown:
+      self.infoMessage(
+        icon: "person.crop.circle.badge.questionmark",
+        label: l10n.State.Unknown.label,
+        secondaryLabel: l10n.State.Unknown.sublabel,
+        color: .orange
+      )
+    case let .existing(name):
+      self.infoMessage(
+        icon: "person.crop.circle.badge.checkmark",
+        label: l10n.State.Existing.label(name),
+        secondaryLabel: l10n.State.Existing.sublabel(name),
+        color: Colors.State.green.color
+      )
+    }
+  }
+
+  static func infoMessage(
+    icon: String,
+    label: String,
+    secondaryLabel: String? = nil,
+    color: Color?
+  ) -> some View {
+    Label {
+      VStack(alignment: .leading) {
+        Text(label.asMarkdown)
+        secondaryLabel.map { Text($0.asMarkdown) }
+          .foregroundColor(.secondary)
+      }
+    } icon: {
+      Image(systemName: icon)
+    }
+    .foregroundColor(color)
+    .frame(maxWidth: .infinity, alignment: .leading)
+  }
+}
+
+// MARK: - The Composable Architecture
+
+// MARK: Reducer
+
+public let addMemberReducer = Reducer<
+  AddMemberSheetState,
+  AddMemberSheetAction,
+  AddMemberSheetEnvironment
+> { state, action, environment in
+  switch action {
+  case let .didUpdateText(text):
+    state.isProcessing = true
+    return Effect(value: .didCheckText(text))
+      .delay(for: 0.5, scheduler: environment.mainQueue)
+      .eraseToEffect()
+
+  case let .didCheckText(text):
+    state.isProcessing = false
+
+    switch text {
+    case "":
+      state.info = .none
+    case "invalid":
+      state.info = .invalid
+    case "unknown@prose.org":
+      state.info = .unknown
+    case let text where text.hasSuffix("."):
+      state.info = .invalid
+    case let text where text.contains("@") && text.contains("."):
+      state.info = .existing("Firstname Lastname")
+    default:
+      state.info = .invalid
+    }
+
+    return .none
+
+  case .binding(\.$text):
+    struct Token: Hashable {}
+    return Effect(value: .didUpdateText(state.text))
+      .debounce(id: Token(), for: 0.5, scheduler: environment.mainQueue)
+      .eraseToEffect()
+
+  default:
+    return .none
+  }
+}.binding()
+
+// MARK: State
+
+public struct AddMemberSheetState: Equatable {
+  public enum InfoMessage: Equatable {
+    case none
+    case invalid
+    case unknown
+    case existing(String)
+  }
+
+  var info: InfoMessage
+  var isProcessing: Bool
+
+  @BindableState var text: String
+
+  var isFormValid: Bool {
+    switch self.info {
+    case .existing, .unknown:
+      return true
+    default:
+      return false
+    }
+  }
+
+  public init(
+    text: String = "",
+    info: InfoMessage = .none,
+    isProcessing: Bool = false
```

Comment:
Don't inject internal state from the outside (`isProcessing` and `info`). `text` could be used for editing purposes, albeit that will change to a JID in the future.